pkg/storage/cacher.go

```diff
@@ -464,7 +467,15 @@ func (c *cacheWatcher) stop() {
 }
 
 func (c *cacheWatcher) add(event watchCacheEvent) {
-	c.input <- event
+	select {
+	case c.input <- event:
+	case <-time.After(time.Second):
```

Comment:
I'm concerned that this will start randomly disconnecting watchers when under heavy load (e.g. in tests). @brendandburns shall we merge anyway? I think it's better to prevent the deadlock.